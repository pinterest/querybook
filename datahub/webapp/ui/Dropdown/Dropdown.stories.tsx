import React from 'react';
// eslint-disable-next-line import/no-extraneous-dependencies
import centered from '@storybook/addon-centered/react';

import { Dropdown } from './Dropdown';
import { ListMenu } from 'ui/Menu/ListMenu';
import { Button } from 'ui/Button/Button';

export default {
    title: 'Stateful/Dropdown',
    decorators: [centered],
};
export const _Dropdown = (args) => (
    <>
        <Dropdown {...args}>
            <ListMenu
                items={[{ name: 'Dropdown Item' }, { name: 'with ListMenu' }]}
            />
        </Dropdown>
        <div className="mb12" />
        <Dropdown
            {...args}
            customButtonRenderer={() => (
                <Button icon={'heart'} title="Dropdown" />
            )}
        >
            <ListMenu
                items={[
                    { name: 'Soft Dropdown' },
                    { name: 'with Custom Button' },
                    {
                        name: 'Nested Item',
                        items: [
                            { name: 'Nested Option 1' },
                            { name: 'Nested Option 2' },
                        ],
                    },
                ]}
                soft
            />
        </Dropdown>
    </>
);

_Dropdown.args = {
    menuIcon: 'menu',
    hoverable: false,
    isRight: false,
    isUp: false,
};
