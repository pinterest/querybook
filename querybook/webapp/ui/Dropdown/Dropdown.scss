.Dropdown {
    position: relative;
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: flex-start;

    .Dropdown-trigger {
        cursor: pointer;
        display: flex;
        justify-content: center;
    }

    .Dropdown-menu {
        min-width: 80px;
        position: absolute;
        z-index: 20;
        box-shadow: var(--box-shadow-md);
        border-radius: var(--border-radius-sm);
        font-size: var(--small-text-size);
    }

    &.opens-bottom {
        > .Dropdown-menu {
            top: 100%;
        }
    }

    &.opens-top {
        > .Dropdown-menu {
            bottom: 100%;
            padding-bottom: 4px;
            padding-top: initial;
            top: auto;
        }
    }

    &.opens-top,
    &.opens-bottom {
        &.aligns-right {
            > .Dropdown-menu {
                left: auto;
                right: 0;
            }
        }

        &.aligns-left {
            > .Dropdown-menu {
                left: 0;
            }
        }
    }

    &.opens-right {
        > .Dropdown-menu {
            left: 100%;
        }
    }

    &.opens-left {
        > .Dropdown-menu {
            right: 100%;
        }
    }

    &.opens-right,
    &.opens-left {
        &.aligns-top {
            > .Dropdown-menu {
                top: 0;
            }
        }

        &.aligns-bottom {
            > .Dropdown-menu {
                bottom: 0;
            }
        }
    }

    &.Dropdown-nested-menu {
        .Dropdown-menu {
            display: flex;
            top: 0px;
        }
        &:not(.nested-right) {
            .Dropdown-menu {
                &::after {
                    content: ' ';
                    width: 4px;
                }
            }
        }

        .nested-right {
            .Dropdown-menu {
                &::before {
                    content: ' ';
                    width: 4px;
                }
            }
        }
    }
}
