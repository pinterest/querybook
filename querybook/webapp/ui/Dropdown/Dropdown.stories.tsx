import centered from '@storybook/addon-centered/react';
import React from 'react';

import { Button } from 'ui/Button/Button';
import { ListMenu } from 'ui/Menu/ListMenu';

import { Dropdown } from './Dropdown';

export default {
    title: 'Stateful/Dropdown',
    decorators: [centered],
};
export const _Dropdown = (args) => (
    <>
        <Dropdown {...args}>
            <ListMenu
                items={[{ name: 'Dropdown Item' }, { name: 'with ListMenu' }]}
            />
        </Dropdown>
        <div className="mb12" />
        <Dropdown
            {...args}
            customButtonRenderer={() => (
                <Button icon={'ChevronDown'} title="Dropdown" />
            )}
        >
            <ListMenu
                items={[
                    { name: 'Soft Dropdown' },
                    { name: 'with Custom Button' },
                    {
                        name: 'Nested Item',
                        items: [
                            { name: 'Nested Option 1' },
                            { name: 'Nested Option 2' },
                        ],
                    },
                ]}
                soft
            />
        </Dropdown>
    </>
);

_Dropdown.args = {
    menuIcon: 'menu',
    hoverable: false,
    layout: ['bottom', 'right'],
};
