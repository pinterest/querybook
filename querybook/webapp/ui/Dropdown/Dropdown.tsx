import clsx from 'clsx';
import React from 'react';

import { useEvent } from 'hooks/useEvent';
import { IconButton } from 'ui/Button/IconButton';
import type { AllLucideIconNames } from 'ui/Icon/LucideIcons';
import { Popover, PopoverLayout } from 'ui/Popover/Popover';
import { AccentText } from 'ui/StyledText/StyledText';

import './Dropdown.scss';

const defaultLayout: PopoverLayout = [
    'bottom', // main axis, the menu opens from the bottom side
    'left', // off axis, determines if menu aligns right or left
];
interface IProps {
    menuIcon?: AllLucideIconNames;
    className?: string;
    customButtonRenderer?: () => React.ReactNode;

    hoverable?: boolean;
    layout?: PopoverLayout;

    usePortal?: boolean;
}

export const Dropdown: React.FunctionComponent<IProps> = ({
    menuIcon = 'Menu',
    className,
    customButtonRenderer,

    hoverable = true,
    layout = defaultLayout,
    usePortal,

    children,
}) => {
    const selfRef = React.useRef<HTMLDivElement>(null);
    const popoverRef = React.useRef<HTMLDivElement>(null);

    const [active, setActive] = React.useState(false);

    // Hover Based Dropdown control
    const handleMouseEnter = React.useCallback(() => {
        if (hoverable) {
            setActive(true);
        }
    }, [hoverable]);
    const handleMouseExit = React.useCallback(() => {
        if (hoverable) {
            setActive(false);
        }
    }, [hoverable]);

    // Click based Dropdown Control
    const handleClick = React.useCallback((event) => {
        if (!hoverable) {
            event.stopPropagation();
            event.preventDefault();
            setActive(true);
        }
    }, []);
    const onDocumentClick = React.useCallback(
        (event: MouseEvent) => {
            if (
                !(
                    event.composedPath().includes(selfRef.current) ||
                    (popoverRef.current &&
                        event.composedPath().includes(popoverRef.current))
                )
            ) {
                setActive(false);
            }
        },
        [selfRef.current]
    );
    useEvent('mousedown', onDocumentClick, {
        element: document,
        disabled: hoverable || !active,
    });

    const buttonDOM = customButtonRenderer ? (
        customButtonRenderer()
    ) : (
        <IconButton icon={menuIcon} noPadding />
    );

    const customFormatClass = customButtonRenderer ? ' custom-format ' : '';

    let dropdownContent =
        active && children ? (
            <div className="Dropdown-menu" role="menu">
                {children}
            </div>
        ) : null;

    if (usePortal && dropdownContent) {
        dropdownContent = (
            <Popover
                ref={popoverRef}
                onHide={() => setActive(false)}
                anchor={selfRef.current}
                layout={layout}
                hideArrow
                skipAnimation
                noPadding
            >
                {dropdownContent}
            </Popover>
        );
    }

    const mainAxis = layout[0];
    const offAxis =
        layout.length > 1
            ? layout[1]
            : mainAxis === 'left' || mainAxis === 'right'
            ? 'top'
            : 'right';
    const combinedClassName = clsx({
        Dropdown: true,
        [className]: className,
        [customFormatClass]: true,
        [`opens-${mainAxis}`]: !usePortal,
        [`aligns-${offAxis}`]: !usePortal,
    });

    return (
        <div
            className={combinedClassName}
            onMouseEnter={handleMouseEnter}
            onMouseLeave={handleMouseExit}
            onClick={handleClick}
            ref={selfRef}
        >
            <AccentText className="Dropdown-trigger" weight="bold">
                {buttonDOM}
            </AccentText>
            {dropdownContent}
        </div>
    );
};
