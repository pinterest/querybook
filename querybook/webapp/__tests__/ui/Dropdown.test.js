import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import React from 'react';
import renderer from 'react-test-renderer';

import { Dropdown } from '../../ui/Dropdown/Dropdown';

it('renders without crashing', () => {
    shallow(<Dropdown />);
});

describe('matches enzyme snapshots', () => {
    it('matches snapshot', () => {
        const wrapper = shallow(<Dropdown />);
        const serialized = toJson(wrapper);
        expect(serialized).toMatchSnapshot();
    });
});

describe('matches test renderer snapshot', () => {
    it('serializes the styles', () => {
        const output = renderer.create(<Dropdown />);
        expect(output).toMatchSnapshot();
    });
});
