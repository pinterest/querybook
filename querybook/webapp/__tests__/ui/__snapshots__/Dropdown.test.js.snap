// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches enzyme snapshots matches snapshot 1`] = `
<div
  className="Dropdown  opens-bottom aligns-left"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <AccentText
    className="Dropdown-trigger"
    weight="bold"
  >
    <ForwardRef
      icon="Menu"
      noPadding={true}
    />
  </AccentText>
</div>
`;

exports[`matches test renderer snapshot serializes the styles 1`] = `
<div
  className="Dropdown  opens-bottom aligns-left"
  onClick={[Function]}
  onMouseEnter={[Function]}
  onMouseLeave={[Function]}
>
  <div
    className="StyledText__StyledTextDiv-sc-1ldibmc-0 lcQBiO Text Dropdown-trigger"
  >
    <span
      className="IconButton no-padding"
    >
      <span
        className="Icon"
      >
        <svg
          className="lucide lucide-menu"
          fill="none"
          height={24}
          stroke="currentColor"
          strokeLinecap="round"
          strokeLinejoin="round"
          strokeWidth={2}
          viewBox="0 0 24 24"
          width={24}
          xmlns="http://www.w3.org/2000/svg"
        >
          <line
            x1="4"
            x2="20"
            y1="12"
            y2="12"
          />
          <line
            x1="4"
            x2="20"
            y1="6"
            y2="6"
          />
          <line
            x1="4"
            x2="20"
            y1="18"
            y2="18"
          />
        </svg>
      </span>
       
    </span>
  </div>
</div>
`;
